:memo: consistency improvements on SNIP-29

Frostweeds opened the pull request:
# Context

Based on observations we made during our rewriting of the Paymaster internally for open-sourcing as well as discussion we had both we Starkware and Partners we would like to propose the following changes to the specification.

## Rational

We believe the current specification does not properly represent the semantic of the paymaster and we think it might negatively impact DevX and slow down integrations. More specifically, we believe the following points that we illustrate with some examples.

1. The current specification leaks too much information about the internals of the paymaster.

Right now we expect the developer to call `build` and then call `execute`. Suppose that the developer calls `build` with some deployment data and some call. The response of `build` contains the typed data that is composed of the call + the transfer of the gas token so this doesn't include the deployment data. Now to execute that call the developer has to send the typed data + the deployment data again and a signature on the typed data.

We believe that the optional nature of the "deployment_data" doesn't reflect the "Deploy" semantic associated to it. It also makes the support of "Deploy-Only" transaction not natural (potentially requiring a different endpoint). It might also be possible, assuming the invoke doesn't rely on the newly deployed contract, that calling `execute` only with the typed data might result in a valid execution.

2. The specification relies too much on the developers to figure out what are the constraints of the system. 

Currently we only support transaction paying in a different token. Nonetheless, we know that there is a need for sponsored transaction which is why `gas_token` in an optional parameters. On a recent discussion with Starkware, we also mentioned the support of "tip" which was also proposed an an optional parameters.

We believe this design will put a lot of complexity on the developers and allow invalid behaviors (ex. sponsoring + tip).

3. The design isn't oriented towards the future as it doesn't leave room for upgrades.

This point echo some the observation we've already made which is that right now parameters are either required or represented as optional and depending on which one is set or not we have different semantic. This makes it very difficult to add additional parameters or execution mode in a sound manner.

## Proposals

We propose to introduce proper typing that reflect the expect semantic of the endpoints and explicitly expose the constraints of the system.

More specifically we propose first, to differentiate explicitly the transaction, i.e to have 3 modes
```
Deploy { /* deployment_data */},
Invoke{ /* invoke_data */}
DeployAndInvoke{ /* deployment_data + invoke_data */ }
```

This already avoid relying on the optionality of the `deployment_data` to fix the semantic of the build / execute. Furthermore, it emphasis the intended behavior

The second modification is the introduction of versioned execution parameters that can easily be extended potentially avoiding an entire new API version. We explicit the construct using Pseudo-Rust for the purpose of the discussion

```
enum ExecutionParameters {
     V1 { // Today
         fee_mode: FeeMode,
         time_bounds: Option<TimeBounds>
     },
}

enum FeeMode {
     Sponsored,
     Default { gas_token: Felt },
     Priority { gas_token: Felt, tip: Felt }
}
```

The idea here is to extend ExecutionParameters with new version when the changes only require a minor version bump while we can proceed with path versioning on major version bump.

leo-starkware commented on assets/snip-29/paymaster_api.json:
```diff
@@ -222,76 +209,329 @@
-            "errors": [
-                {
-                    "$ref": "#/components/errors/INVALID_ID"
+            "PAYMASTER_EXECUTION": {
```
Suggest: USER_PARAMETERS

Frostweeds replied:
Fixed

leo-starkware commented on assets/snip-29/paymaster_api.json:
```diff
@@ -222,76 +209,329 @@
+                    }
+                ]
+            },
+            "PAYMASTER_INVOKE": {
```
Suggest: USER_INVOKE

Frostweeds replied:
Fixed

leo-starkware commented on assets/snip-29/paymaster_api.json:
```diff
@@ -222,76 +209,329 @@
-                                "dropped"
-                            ]
+            },
+            "PAYMASTER_TRANSACTION": {
```
Suggest: USER_TRANSACTION

leo-starkware commented on assets/snip-29/paymaster_api.json:
```diff
@@ -222,76 +209,329 @@
+                    }
+                ]
+            },
+            "PAYMASTER_FEE_MODE": {
```
Suggest: FEE_MODE (and delete the currently existing FEE_MODE objects).

leo-starkware commented on assets/snip-29/paymaster_api.json:
```diff
@@ -222,76 +209,329 @@
+                    }
+                ]
+            },
+            "PAYMASTER_FEE_ESTIMATE": {
```
Suggest: FEE_ESTIMATE

leo-starkware commented on assets/snip-29/paymaster_api.json:
```diff
@@ -222,76 +209,329 @@
+                        "properties": {
+                            "type": {
+                                "title": "Transaction type",
+                                "const": [
```
change all the "const" to "enum"

Frostweeds replied:
Did the change but what's the reason ?

leo-starkware commented on assets/snip-29/paymaster_api.json:
```diff
@@ -222,76 +209,329 @@
+                            "gas_token": {
+                                "$ref": "#/components/schemas/FELT"
+                            },
+                            "tip": {
```
In what units is the tip? I would suggest strk, since people will look onchain and see what tip should be regarded as "high/low priority"

Frostweeds replied:
Yes we believe it should be in strk as well. I changed the name to `tip_in_strk`.

leo-starkware commented on assets/snip-29/paymaster_api.json:
```diff
@@ -140,39 +154,12 @@
-                    "description": "The address of the user account",
+                    "name": "transaction",
                     "required": true,
                     "schema": {
```
Shouldn't there be an optional tip? Either as an optional param of the request or as an optional field in PAYMASTER_EXECUTABLE_TRANSACTION?

leo-starkware commented on assets/snip-29/paymaster_api.json:
```diff
@@ -222,76 +209,329 @@
+                "description": "Execution parameters to used when executing the transaction through the paymaster",
+                "oneOf": [
+                    {
+                        "title": "Execution parameters v1",
```
If I understand correctly, this is the versioning of the spec object, and it's uncoupled from the versioning of the outside execution objects. Maybe worth mentioning here

Frostweeds replied:
Yes exactly, I added some precision, let me know if it's enough

leo-starkware reviewed (commented):
1. (naming) I think that too many objects are prefixed with `PAYMASTER_` and it makes the API less readable. I made suggestions for: `PAYMASTER_TRANSACTION`, `PAYMASTER_EXECUTION`, `PAYMASTER_INVOKE`, `PAYMASTER_FEE_MODE`, `PAYMASTER_FEE_ESTIMATE`.
2. Delete the previously existing objects `OUTSIDE_EXECUTION_TYPED_DATA_V3`, `OUTSIDE_EXECUTION_MESSAGE_V3`, `FEE_MODE`.
3. Change the "const" keywords into "enum"
4. Tip in `execute`

Frostweeds commented:
@leo-starkware thanks a lot for your review !

I've made most of the changes you've suggested, however I don't know how what you means by removing the object `FEE_ESTIMATE`, `EXECUTE_FROM_OUTSIDE`, ... We are referencing it and if I remove them then that will break the spec, no ?What did you have in mind ?

On another note we decided to propose a second round of changes, mostly on naming and also the payload returned by the build. More specifically

- `getSupportedTokensWithPrice` -> `getSupportedTokens`

The main purpose of this method if to return the supported tokens so we modify the name to reflect that.

- `buildTypedData` -> `buildTransaction` & `execute` -> `executeTransaction`

We decided to standardize the naming because we believe that it better reflects the intended flow which is user calling build then execute while we think the previous name hinted at a fragmented process.

- Send the deployment data and user parameters in the build response

Here the rational is to leave no room for misinterpretation in the expect user flow (build -> execute). Before the user had to potentially take some data from the build request and some data from the build response to send to execute which does not feel natural. With this new spec, everything that the user needs for execution is coming from the build response which we think is closed to the expected behavior of a build endpoint.

Additionally, as we implemented the deploy-only transaction and the sponsored transaction, we've remarked that we have scenario where the previous response didn't make sense as can be seen in the following 2 cases

Case 1: Deploy-Only not Sponsored: Here the build must be called because even if we are only performing a deploy, it will actually yield some typed data containing the gas token transfer so the resulting will actually be a Deploy & Invoke.

Case 2: Deploy-Only Sponsored: Here the build will not create any typed data and only perform and estimate so the previous response of the build didn't support that case which would mean that we would raise an error.

We believe that this leaks the internals of the paymaster and puts the burden on the user to know when to call build + execute or just execute which we believe is problematic. With the proposed modification the user always follow the same flow which is build -> execute no matter the transaction and parameters.

leo-starkware commented:
I agree with the modifications to the `buildTransaction` response - it adds some redundancy but should improve UX.

IMO the only change to be done is to remove:
- OUTSIDE_EXECUTION_TYPED_DATA_V3
- OUTSIDE_EXECUTION_MESSAGE_V3
- Remove the reference to OUTSIDE_EXECUTION_TYPED_DATA_V3 in OUTSIDE_EXECUTION_TYPED_DATA
(this is what I meant in the previous review)

Re the change to `getSupportedTokensWithPrice` - I think the new name is fine as welll, but note that the object still contains price data

Frostweeds commented:
Perfect ! I've removed the components you've mentioned plus FEE_ESTIMATE and renamed PAYMASTER_FEE_ESTIMATE into FEE_ESTIMATE as you've proposed in your first review.

Regarding the `getSupportedTokens`, yes we decided to keep the price in it but as convenience instead of tying it to the actual semantic of the endpoint. Is that ok ?

Let me know if there's anything that needs to be changed, otherwise I believe we can move forward.

leo-starkware commented on assets/snip-29/paymaster_api.json:
```diff
@@ -136,43 +215,24 @@
-                        "title": "Typed data",
-                        "$ref": "#/components/schemas/OUTSIDE_EXECUTION_TYPED_DATA"
+                        "title": "Transaction",
+                        "description": "Typed data build by calling paymaster_buildTypedData signed by the user to be executed by the paymaster service",
```
Change to "paymaster_buildTransaction"
